feature(forms): moves datepicker init to AMD and improves dev usability

hypeJunction opened the pull request:
Datepicker initialization now takes place in an AMD module on per input basis, which allows
plugins to configure datepicker options by passing datepicker_options to the input view.
- [x] Add a note that `onSelect` now uses the datepicker instance values to parse the timestamp value, which allows us to modify the `dateFormat` option without breaking the parsing
- [x] address https://github.com/Elgg/Elgg/pull/9224#discussion_r48444274
- [x] the inline datepicker shouldn't display the full width of the viewport, right?

hypeJunction committed:
feature(forms): moves datepicker init to AMD and improves dev usability

Datepicker initialization now takes place in an AMD module on per input basis, which allows
plugin to configure datepicker options by passing datepicker_options to the input view.

hypeJunction commented on js/lib/ui.js:
```diff
@@ -326,38 +326,7 @@ elgg.ui.loginHandler = function(hook, type, params, options) {
-		deps.push('jquery-ui/i18n/datepicker-'+ elgg.get_language() + '.min');
-	}
-	require(deps, loadDatePicker);
+	elgg.deprecated_notice('elgg.ui.initDatePicker has been deprecated. Use input/date AMD module instead');
```
We can actually require the AMD module here, and make this PR fully BC, by initializing the datepicker on `.elgg-input-date`, in case someone calls this function directly.

mrclay replied:
For BC we can't replace synchronously defined functions with async modules, right?

hypeJunction replied:
You saying it's not possible, or do you mean that synchronicity is part of
BC?
On Dec 15, 2015 1:10 AM, "Steve Clay" <notifications@github.com> wrote:

> In js/lib/ui.js
> <https://github.com/Elgg/Elgg/pull/9224#discussion_r47579776>:
>
> > -			changeMonth: true,
> > -			changeYear: true
> > -		});
> > -	}
> > -
> > -	if (!$('.elgg-input-date').length) {
> > -		return;
> > -	}
> > -
> > -	// require language module if necessary
> > -	var deps = [];
> > -	if (elgg.get_language() != 'en') {
> > -		deps.push('jquery-ui/i18n/datepicker-'+ elgg.get_language() + '.min');
> > -	}
> > -	require(deps, loadDatePicker);
> > +	elgg.deprecated_notice('elgg.ui.initDatePicker has been deprecated. Use input/date AMD module instead');
>
> For BC we can't replace synchronously defined functions with async
> modules, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9224/files#r47579776>.
>

mrclay replied:
eh, ignore my comment for now.

mrclay replied:
> We can actually require the AMD module here

Yes, this should work. Sorry for the noise.

hypeJunction commented:
This should be BC now.

mrclay commented on views/default/input/date.js:
```diff
@@ -0,0 +1,62 @@
+
+	if (elgg.get_language() !== 'en') {
+		// require language module if necessary
+		require('jquery-ui/i18n/datepicker-' + elgg.get_language() + '.min');
```
This can't work. 1. It would need to be `require([' ... '])` to load async. 2. We have to guarantee the language module is loaded before `$elem.datepicker()` is called.

I think I have a solution I can post tomorrow. Basically we'll use promises to delay the `$elem.datepicker()` call until the language is loaded in an async `require` call.

hypeJunction replied:
I thought `define(function(require) {});` pattern is equivalent to async `define([], function() {})`

mrclay replied:
Within a define function RequireJS uses `Function.toString()` trickery to allow synchronous-looking code like `require('foo')`, but it's not really synchronous. The `'foo'` is discovered and loaded before the define function is called.

mrclay commented on views/default/input/date.php:
```diff
@@ -48,8 +54,21 @@
+?>
+<script>
+	require(['input/date'], function(datepicker) {
+		datepicker.init('<?= $selector ?>');
```
Note this might result in a broken string:
```js
datepicker.init('.elgg-input-date[name='myname']');
```
Solution is to always use json_encode() to write JS string literals.

mrclay commented on mod/developers/views/default/theme_sandbox/forms.php:
```diff
@@ -103,6 +103,22 @@
+			'value' => '1-Jan-2016',
+			'timestamp' => true,
+			'datepicker_options' => array(
+				'dateFormat' => 'd-M-yy',
```
For some reason this limits me to 2010 June 11 to to July 11. Looking into it.

mrclay commented on mod/developers/views/default/theme_sandbox/forms.php:
```diff
@@ -103,6 +103,22 @@
+				'dateFormat' => 'd-M-yy',
+				'changeMonth' => false,
+				'changeYear' => false,
+				'minDate' => '1-Jan-2016',
```
These short month names have to match the language that datepicker uses. So numbers work best. PR coming.

mrclay commented:
I made a PR against your branch. https://github.com/hypeJunction/Elgg/pull/2

hypeJunction commented:
Just merged it.

hypeJunction commented:
I will update the inline docs and squash, when I am back at my computer.

mrclay commented on js/lib/ui.js:
```diff
@@ -467,6 +442,5 @@ elgg.ui.initAccessInputs = function () {
 };
 
 elgg.register_hook_handler('init', 'system', elgg.ui.init);
-elgg.register_hook_handler('init', 'system', elgg.ui.initDatePicker);
```
This would stop the auto-wiring of .elgg-input-date inputs not using the input/date view. For BC we really have to catch them all.

hypeJunction replied:
We can't throw deprecation errors if we gonna rewire them. Alternative to add an extra selector/data attr in input/date view, and log a deprecation warning if it's not there.

hypeJunction commented:
@mrclay Take a look. I have updates the AMD module to iterate through all selector instances. `data-datepicker-opts` is now added to all inputs rendered with `input/date`, which allows us to catch elements rendered outside of the view. I have added an inline datepicker to theme sandbox to test BC.

mrclay commented on views/default/input/date.php:
```diff
@@ -48,8 +52,20 @@
 	$vars['value'] = gmdate('Y-m-d', $vars['value']);
 }
 
+$datepicker_options = elgg_extract('datepicker_options', $vars);
```
Just make the default an empty string here. Then you don't need the ternary: just json_encode either way.

hypeJunction replied:
I am afraid if someone passes `false` we would end up with problems in JS `$.extend`

mrclay replied:
Actually default should probably be `new stdClass`, since $.extend is going to read from it.

hypeJunction replied:
IMO, an empty string is better, as JS will interpret it as null and set to empty object.

mrclay replied:
Only other thing we probably want is to move logic to elgg.ui_initDatePicker (`@private` in jsdoc), call that from core. elgg.ui.initDatePicker would just show deprecation notice and call the private one.

mrclay replied:
On 2nd thought that can wait. We need a big strategy to deprecate most of these instead of one by one

hypeJunction replied:
Well, if we manage to deprecate them all before 3.x, we can just get rid of
them without any new APIs. It will be a challenge.
On Dec 18, 2015 12:47 AM, "Steve Clay" <notifications@github.com> wrote:

> In views/default/input/date.php
> <https://github.com/Elgg/Elgg/pull/9224#discussion_r47978137>:
>
> > @@ -48,8 +52,20 @@
> >  	$vars['value'] = gmdate('Y-m-d', $vars['value']);
> >  }
> >
> > +$datepicker_options = elgg_extract('datepicker_options', $vars);
>
> On 2nd thought that can wait. We need a big strategy to deprecate most of
> these instead of one by one
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9224/files#r47978137>.
>

hypeJunction commented:
@mrclay Anything holding this back? I suppose you are considering other options for updating all the JS, but I wouldn't want this to go stale.

mrclay commented:
I get deprecation notices on the sandbox theme forms.

mrclay commented:
Also the inline datepicker os the full width of the viewport.

mrclay commented on mod/developers/views/default/theme_sandbox/forms.php:
```diff
@@ -103,6 +103,32 @@
+				'class' => 'elgg-input-date',
+			)),
+			'label' => elgg_view('elements/forms/label', array(
+				'label' => 'Inine datepicker (div.elgg-input-date)',
```
"Inline". I'm not sure we should encourage devs to duplicate this usage (no input element). And without the `data-` attribute it causes the deprecation notice.

hypeJunction commented:
@mrclay Addressed your comments

mrclay commented on views/default/elements/forms.css.php:
```diff
@@ -302,6 +302,7 @@
 }
 .ui-datepicker-inline {
 	box-shadow: none;
+	max-width: 225px;
```
I wonder why it worked before without this

mrclay commented:
Great!
